doc/developer/design/20240522_mir_representation_types.md

```diff
@@ -0,0 +1,133 @@
+
+1. Rename the existing `ScalarType` to `SqlScalarType`. (We may want to rename `ColumnType` and `RelationType`, as well.)
+2. Have MIR work with `ReprScalarType` (see [MVP](#minimal-viable-prototype)).
+   a. Introduce the `ReprScalarType` datatype (see [MVP](#minimal-viable-prototype)).
+   b. Change the optimizer to work with `ReprScalarType` throughout.
+   c. Change the adapter to record the `SqlScalarType` of a query and update its nullability information using the `ReprScalarType` at the end of optimization.
+3. Elide noop casts, confirming that we plan the problematic joins better.
+
+PR #2 will have a large diff.
+On the plus side, it should be a largely mechanical change: there is a straightforward way to project a `SqlScalarType` to a `ReprScalarType`, and most uses of types in the optimizer will be parametric.
+
+## Minimal Viable Prototype
+
+The representation types will look like the following:
+
+```rust
+pub enum ReprScalarType {
+  Bool,
+  Int16,
+  Int32,
+  Int64,
+  UInt8, // also includes SqlScalarType::PgLegacyChar
+  UInt16,
+  UInt32, // also includes SqlScalarType::{Oid,RegClass,RegProc,RegType}
+  UInt64,
+  Float32,
+  Float64,
+  Numeric,
+  Date,
+  Time,
+  Timestamp  { precision: Option<TimestampPrecision>, },
+  TimestampTz { precision: Option<TimestampPrecision>, },
+  MzTimestamp,
+  Interval,
+  Bytes,
+  Jsonb,
+  String, // also includes SqlScalarType::{VarChar,Char,PgLegacyName}
+  Uuid,
+  Array(Box<ReprScalarType>), // also includes SqlScalarType::Int2Vector
+  List { element_type: Box<ReprScalarType> }, // also includes SqlScalarType::Record
+  Map { value_type: Box<ReprScalarType> },
+  Range { element_type: Box<ReprScalarType> },
+  MzAclItem,
+  AclItem,
+}
+
+pub struct ReprColumnType {
+    pub scalar_type: ReprScalarType,
+    pub nullable: bool,
+}
+```
+
+We'll need to define a projection from `SqlScalarType` to `ReprScalarType` (see [`ScalarType::is_instance_of()`](https://github.com/MaterializeInc/materialize/blob/e4578164fb28a204b43c58ab8ff6c1d3e3b4427f/src/repr/src/scalar.rs#L947) for the correspondence).
+
+## Alternatives
+
+We considered several alternatives:
+
+- (`eq-type`) Change the MIR typechecker to conflate `varchar` and `text` types (and drop calls to `varchar_to_text`).
+- (`eq-indx`) Change index selection to conflate `varchar` and `text` types (and leave calls to `varchar_to_text`).
+- (`remixvc`) Automatically turn add `varchar_to_text` to all indices on a `varchar`.
+- (`remix++`) Automatically wrap every use of a `varchar` with `varchar_to_text`.
+
+The `eq-type` and `eq-indx` approach came up in early discussions.
+These are fundamentally hacks, with the upside of being quick fixes.
+The downside is the addition of a tricky corner case, some risk (what if keeping the types separate matters somewhere we've forgotten about?), and some technical debt/mess-making.
```

Comment:
What makes them fundamentally hacks/tech debt? `eq-indx` in particular seems like a narrow case of a smart index selector (let's call it `conv-indx`) that can handle mixed type lookups, e.g. looking up an i32 in an i64 index or vice-versa, and that seems useful/powerful/desirable?

Reply:
The hacky/tech-debt/messy feeling here is that it's a very specialized way to work with `ScalarExpr` in a very narrow context, and I worry about duplicated work and inconsistent handling.

But your point about numeric indices makes it clear to me that index selection will need to be a _little bit_ special no matter what---and index creation, too (#27030).

So, I'm coming around to `eq-indx`/`conv-indx`. An advantage of `mirtype` is that it will let us remove many infallible casts wholesale while simplifying our view of types at the MIR level, which will address our immediate problem. But we'll need something like `conv-indx` in general. (With `conv-indx`, you could avoid doing the work of #15476. Or, alternatively, it may be possible to implement `conv-indx` using the techniques from #15476.)

Reply:
Something that's been nagging at me, but I may simply be confused about: the hash functions for `i32` and `i64` are not going to be the same, i.e., sign extending `n::i32` to `signext(n)::i64` will likely not hash the same, so I think we'd need some careful logic here. I remember bringing something similar up to @frankmcsherry (I think using an index on `n::i64` to lookup `n+1` or some similarly invertible operation) and him pointing out why it wouldn't work, but I can't remember what the exact issue was.

Reply:
I'm not really sure how `eq-indx` would work in a join when the actual representations are different. If we have an index on `t1` with a `Datum::Int32` key, and an index on `t2` with a `Datum::Int64` key, then at what moment could we do the cast? I think our current join rendering just does exact equality on the arrangement keys as they are, and there is currently no way to give it a closure or something which would sneakily convert between `Int32` and `Int64`.

So, in addition to modifying the index selection in `JoinImplementation`, we'd need to somehow also modify the join rendering (for Differential join it's `mz_join_core.rs`, and for Delta join it's `dogsdogsdogs::operators::half_join::half_join_internal_unsafe`, in the DD repository). This should involve a solution to the hashing problem mentioned above, besides changing how the keys are equated (somewhere around the `seek_key` calls).

(Note that this problem didn't occur in https://github.com/MaterializeInc/materialize/pull/15476, because there it's only one side of the equality that is doing an index lookup, so there we just move the fiddling with the types to the other side of the equality.)

Reply:
We discussed this a bit with @mgree, and the hashing issue might be worse than I thought. We are doing partitioned joins, i.e., both of the arrangements are partitioned by a hash, and pairs of rows that the join will emit need to be in the same partition. But now with a cast on one of the keys, the partitioning would need to be modified on one of the arrangements, and I'm not sure if there is any better way to do this than to rebuild the arrangement, which would completely defeat the purpose.

Reply:
Maybe @frankmcsherry and I can talk about how `eq-indx` would work some time this month? I agree that it seems complicated. 😔

Reply:
(Bw., not an ideal solution, but we could add a notice that would make it easier for users to find these implicit cast issues and manually fix them by tweaking their index keys to include the cast: https://github.com/MaterializeInc/materialize/issues/22133)

Reply:
Oh, look what I found!
https://github.com/MaterializeInc/materialize/issues/10871
It's a really well-preserved archeological specimen: haven't been touched or seen in years, and even features the old EXPLAIN output format!

On a more serious note, suddenly finding this long-forgotten, really detailed discussion of these same problems that we are discussing now really demonstrates the cost of the churn of optimizer people...

Reply:
Interesting!

One possibility for `eq-idx` is to automatically coerce to "widest-possible" when doing hashing and ordering. That is, we would hash and order `int16`, `int32`, and so on all as if it were `int128` (and we would similarly hash and order `uint16`, `uint32`, and on on as if it were `uint128`). I don't think there's any storage cost to doing this, just a small computational cost. And then we'd be able to automatically use a wider index for narrower values.

We could even treat everything as `uint128` for hashing purposes, but it probably isn't safe to treat signed ints as `uint128` for ordering purposes (if we ever use ordering to express "real" ordering and not arbitrary ordering).